trace/config.go

```diff
@@ -22,110 +22,160 @@ import (
+// action.
+func NewSpanEndConfig(options ...SpanEndOption) *SpanConfig {
+	c := new(SpanConfig)
+	for _, option := range options {
+		option.applySpanEnd(c)
+	}
+	return c
+}
+
+// SpanOptions applies an option to a SpanConfig. These options are applyable
+// Only when the span is created
 type SpanOption interface {
-	ApplySpan(*SpanConfig)
+	applySpanStart(*SpanConfig)
+}
+
+type SpanOptionFunc func(*SpanConfig)
 
-	// A private method to prevent users implementing the
-	// interface and so future additions to it will not
-	// violate compatibility.
-	private()
+func (fn SpanOptionFunc) applySpanStart(cfg *SpanConfig) {
+	fn(cfg)
+}
+
+type SpanEndOption interface {
+	applySpanEnd(*SpanConfig)
+}
+
+type EventConfig struct {
+	// Attributes describe the associated qualities of a Event.
+	attributes []attribute.KeyValue
+	// Timestamp is a time in a Event was recorded.
+	timestamp time.Time
+}
+
+// Attributes describe the associated qualities of an Event.
+func (cfg *EventConfig) Attributes() []attribute.KeyValue {
+	return cfg.attributes
+}
+
+// Timestamp is a time in an Event life-cycle.
+func (cfg *EventConfig) Timestamp() time.Time {
+	return cfg.timestamp
 }
 
 // NewEventConfig applies all the EventOptions to a returned SpanConfig. If no
 // timestamp option is passed, the returned SpanConfig will have a Timestamp
 // set to the call time, otherwise no validation is performed on the returned
 // SpanConfig.
-func NewEventConfig(options ...EventOption) *SpanConfig {
-	c := new(SpanConfig)
+func NewEventConfig(options ...EventOption) *EventConfig {
+	c := new(EventConfig)
 	for _, option := range options {
-		option.ApplyEvent(c)
+		option.applyEvent(c)
 	}
-	if c.Timestamp.IsZero() {
-		c.Timestamp = time.Now()
+	if c.timestamp.IsZero() {
+		c.timestamp = time.Now()
 	}
 	return c
 }
 
-// EventOption applies span event options to a SpanConfig.
+// EventOption applies span event options to an EventConfig.
 type EventOption interface {
-	ApplyEvent(*SpanConfig)
-
-	// A private method to prevent users implementing the
-	// interface and so future additions to it will not
-	// violate compatibility.
-	private()
+	applyEvent(*EventConfig)
 }
 
-// LifeCycleOption applies span life-cycle options to a SpanConfig. These
-// options set values releated to events in a spans life-cycle like starting,
-// ending, experiencing an error and other user defined notable events.
-type LifeCycleOption interface {
-	SpanOption
-	EventOption
-}
+type attributeOption []attribute.KeyValue
 
-type attributeSpanOption []attribute.KeyValue
-
-func (o attributeSpanOption) ApplySpan(c *SpanConfig)  { o.apply(c) }
-func (o attributeSpanOption) ApplyEvent(c *SpanConfig) { o.apply(c) }
-func (attributeSpanOption) private()                   {}
-func (o attributeSpanOption) apply(c *SpanConfig) {
-	c.Attributes = append(c.Attributes, []attribute.KeyValue(o)...)
+func (o attributeOption) applySpan(c *SpanConfig) {
```

Comment:
is `applySpan` used anymore?  Should attributes be able to be added at span end (and thus implement `applySpanEnd`)?

Reply:
So, our current setup does allow for attributes.  But from our discussion prior the only thing that can be changed at the end is the timeStamp.  

I personally think that attributes should be added at the end as well.  I waffled about if I was to include it, which is why it's in it's half baked state.

Reply:
I think it should be possible to add attributes at span end as well, but I guess it is probably safer to leave that off for now.  We should always be able to add an implementation of `applySpanEnd` later to enable that without breaking any APIs.